File: gtci/09-hashmaps/cpp/2-largest-unique-number.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
   public:
    int largestUniqueNumber(vector<int>&& A) {
        std::unordered_map<int, int> freqMap;

        for (int num : A) freqMap[num] += 1;

        int maxUnique = -1;
        for (int num : A) {
            if (freqMap[num] == 1) maxUnique = std::max(maxUnique, num);
        }

        return maxUnique;
    }
};

int main() {
    Solution sol;

    assert(sol.largestUniqueNumber({5, 7, 3, 9, 4, 9, 8, 3, 2, 5}) == 8);
    assert(sol.largestUniqueNumber({1, 2, 3, 4, 5}) == 5);
    assert(sol.largestUniqueNumber({1, 1, 1, 1}) == -1);

    std::cout << "All test cases passed." << std::endl;

    return 0;
}

File: gtci/09-hashmaps/python3/2-largest-unique-number.py
from typing import List


class Solution:
    def largestUniqueNumber(self, A: List[int]) -> int:
        freq_map = {}

        for num in A:
            if num not in freq_map:
                freq_map[num] = 0
            freq_map[num] += 1

        max_unique = -1
        for num in A:
            if freq_map[num] == 1:
                max_unique = max(max_unique, num)

        return max_unique


if __name__ == "__main__":
    sol = Solution()

    assert sol.largestUniqueNumber([5, 7, 3, 9, 4, 9, 8, 3, 2, 5]) == 8
    assert sol.largestUniqueNumber([1, 2, 3, 4, 5]) == 5
    assert sol.largestUniqueNumber([1, 1, 1, 1]) == -1

    print("All test cases passed.")
